### ISSUE TITLE
Parse logged types

### ISSUE THREAD

**Issue by digorithm** [FuelLabs/fuels-rs#562]
Once https://github.com/FuelLabs/sway/pull/2689 is merged, it will enable the SDK to parse concrete types logged in Sway programs. Which is something we've been long waiting for! 

The SDK should support this by parsing the `loggedTypes` arrays that will show up in the JSON ABI file. 

From the above PR, the relevant points that are going to be needed to implement this support in the SDK:

> Update the JSON ABI generation to also include a loggedTypes array. This is an array of JsonLoggedType which contains a unique ID per log as well as a JsonTypeApplication. Types that are logged are added to the types array which is shared with the contract functions.
> Introduce an IR instruction log which takes a value to log, its type, and an ID (has to be u64). This ID is basically the ID of the log that is emitted in the JSON file

**Comment by mohammadfawaz** [comment]
Another really important point is the following, just to know where to look for the log ID in the `log` instruction itself.
> - Codgen transforms the `log` instruction into either `log` or `logd` depending on the type of the value being logged
>  - In the case of a `log`, `$rA` contains the value being logged and `$rB` contains the ID of the log
>  - In the case of a `logd`, `$rB` contains the ID of the log while `$rC` and `$rD` both describe the value being logged

**Comment by mohammadfawaz** [comment]
One more thing... `forc 0.24.1` disabled the generation of the logged types in the JSON file by default due to an issue with logging (monomorphized) generics (as parsing generic types in not supported yet). Instead, I've exposed the functionality using the `forc build` flag: `--generate-logged-types`. So, when working on this, pass that flag to `forc build` to get the logged types in the JSON file. Once support for this and for generics is complete, we can then remove the flag and turn it on by default.

**Comment by QuinnLee** [comment]
> One more thing... `forc 0.24.1` disabled the generation of the logged types in the JSON file by default due to an issue with logging (monomorphized) generics (as parsing generic types in not supported yet). Instead, I've exposed the functionality using the `forc build` flag: `--generate-logged-types`. So, when working on this, pass that flag to `forc build` to get the logged types in the JSON file. Once support for this and for generics is complete, we can then remove the flag and turn it on by default.

Do the `-flat` json ABI's also have `loggedTypes`

**Comment by digorithm** [comment]
> > One more thing... `forc 0.24.1` disabled the generation of the logged types in the JSON file by default due to an issue with logging (monomorphized) generics (as parsing generic types in not supported yet). Instead, I've exposed the functionality using the `forc build` flag: `--generate-logged-types`. So, when working on this, pass that flag to `forc build` to get the logged types in the JSON file. Once support for this and for generics is complete, we can then remove the flag and turn it on by default.
> 
> 
> 
> Do the `-flat` json ABI's also have `loggedTypes`

There are no more JSON ABI files with "-flat" -- this was removed last week IIRC. The one and only JSON ABI file outputted by the current version of forc is already the flat one and it has the loggedTypes! 😁

**Comment by nfurfaro** [comment]
Somewhat related, but might need its own issue...
When a `require` statement in sway code fails, an arbitrary value is logged by the user, and then there's a `revert(42)`.
The idea being that whenever a `revert` with the value `42` is seen in the logs, it signifies to look at the preceding `log`/`logd` and decode the value there.
